From 2a45b3e8fa0890f514dfd66873d4da91c17c7df0 Mon Sep 17 00:00:00 2001
From: Nicolas Dufresne <nicolas.dufresne@collabora.com>
Date: Fri, 8 Nov 2019 18:00:46 -0500
Subject: [PATCH 32/60] kmssink: Do not drain if imported buffer are from KMS

This adds a check to avoid draining when the imported buffers are in
fact own by kmssink. This happens since we export our kms buffer as
DMABuf. They are not really imported back as we pre-fill the cache,
but uses the same format as if they were external. This fixes
performance issues seen with videocrop2-test (found in -good).

(cherry picked from commit 80b599c81b2e85928ece76d300c2a3075e0f842c)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 sys/kms/gstkmssink.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/sys/kms/gstkmssink.c b/sys/kms/gstkmssink.c
index ce41efc55..69a8c3658 100644
--- a/sys/kms/gstkmssink.c
+++ b/sys/kms/gstkmssink.c
@@ -1712,8 +1712,6 @@ gst_kms_sink_drain (GstKMSSink * self)
 {
   GstParentBufferMeta *parent_meta;
 
-  GST_DEBUG_OBJECT (self, "draining");
-
   if (!self->last_buffer)
     return;
 
@@ -1723,6 +1721,14 @@ gst_kms_sink_drain (GstKMSSink * self)
   if (parent_meta) {
     GstBuffer *dumb_buf, *last_buf;
 
+    /* If this was imported from our dumb buffer pool we can safely skip the
+     * drain */
+    if (parent_meta->buffer->pool &&
+        GST_IS_KMS_BUFFER_POOL (parent_meta->buffer->pool))
+      return;
+
+    GST_DEBUG_OBJECT (self, "draining");
+
     dumb_buf = gst_kms_sink_copy_to_dumb_buffer (self, &self->last_vinfo,
         parent_meta->buffer);
     last_buf = self->last_buffer;
-- 
2.20.1

